apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/extensions/NeutronSecurityGroupExtension.java

```diff
@@ -0,0 +1,382 @@
+                                 input.getProtocol() != null && input.getProtocol().name().equals(ipPermission.getIpProtocol().name()) &&
+                                 input.getPortRangeMin() != null && input.getPortRangeMin() == ipPermission.getFromPort() &&
+                                 input.getPortRangeMax() != null && input.getPortRangeMax() == ipPermission.getToPort();
+                       }
+                    })) {
+               securityGroupApi.deleteRule(rule.getId());
+            }
+         }
+      }
+
+      if (!ipPermission.getGroupIds().isEmpty()) {
+         for (final String groupId : ipPermission.getGroupIds()) {
+            for (Rule rule : filter(securityGroup.getRules(),
+                    new Predicate<Rule>() {
+                       @Override
+                       public boolean apply(@Nullable Rule input) {
+                          return input.getRemoteGroupId() != null && input.getRemoteGroupId().equals(groupId) &&
+                                 input.getProtocol() != null && input.getProtocol().name().equals(ipPermission.getIpProtocol().name()) &&
+                                 input.getPortRangeMin() != null && input.getPortRangeMin() == ipPermission.getFromPort() &&
+                                 input.getPortRangeMax() != null && input.getPortRangeMax() == ipPermission.getToPort();
+                       }
+                    })) {
+               securityGroupApi.deleteRule(rule.getId());
+            }
+         }
+      }
+
+      return getSecurityGroupById(RegionAndId.fromRegionAndId(region, id).slashEncode());
+   }
+
+   @Override
+   public SecurityGroup removeIpPermission(IpProtocol protocol, int startPort, int endPort,
+                                           Multimap<String, String> tenantIdGroupNamePairs,
+                                           Iterable<String> ipRanges,
+                                           Iterable<String> groupIds, SecurityGroup group) {
+      IpPermission.Builder permBuilder = IpPermission.builder();
+      permBuilder.ipProtocol(protocol);
+      permBuilder.fromPort(startPort);
+      permBuilder.toPort(endPort);
+      permBuilder.tenantIdGroupNamePairs(tenantIdGroupNamePairs);
+      permBuilder.cidrBlocks(ipRanges);
+      permBuilder.groupIds(groupIds);
+
+      return removeIpPermission(permBuilder.build(), group);
+   }
+
+   @Override
+   public boolean supportsTenantIdGroupNamePairs() {
+      return false;
+   }
+
+   @Override
+   public boolean supportsTenantIdGroupIdPairs() {
+      return false;
+   }
+
+   @Override
+   public boolean supportsGroupIds() {
+      return true;
+   }
+
+   @Override
+   public boolean supportsPortRangesForGroups() {
+      return false;
+   }
+
+   @Override
+   public boolean supportsExclusionCidrBlocks() {
+      return false;
+   }
+
+   private Function<org.jclouds.openstack.neutron.v2.domain.SecurityGroup, SecurityGroup> toSecurityGroup(final Location location) {
```

Comment:
Move this to its own type.
The private method is a bit awkward. Create a type for this function and use google assisted inject and a factory interface to populate the Location field where needed.